gpiozero/pins/native.py

```diff
@@ -98,52 +100,173 @@ class GPIOFS(object):
+            except KeyError:
                 result = None
                 # Dirty hack to wait for udev to set permissions on
-                # gpioN/direction; there's no other way around this as there's
-                # no synchronous mechanism for setting permissions on sysfs
+                # gpioN/value; there's no other way around this as there's no
+                # synchronous mechanism for setting permissions on sysfs
                 for i in range(10):
                     try:
-                        result = io.open(self.path('gpio%d/value' % pin), 'w+b', buffering=0)
+                        # Must be O_NONBLOCK for use with epoll in edge
+                        # triggered mode
+                        result = os.open(self.path_value(pin),
+                                         os.O_RDONLY | os.O_NONBLOCK)
                     except IOError as e:
                         if e.errno == errno.ENOENT:
-                            with io.open(self.path('export'), 'wb') as f:
+                            with io.open(self.path('export'), 'w+b') as f:
                                 f.write(str(pin).encode('ascii'))
-                            # Pin wasn't exported, so correct the ref-count
-                            self._pin_refs[pin] = 0
                         elif e.errno == errno.EACCES:
                             sleep(i / 100)
                         else:
                             raise
                     else:
+                        self._exports[pin] = result
                         break
-                if not result:
+                # Same for gpioN/edge. It must exist by this point but the
+                # chmod -R may not have reached it yet...
+                for i in range(10):
+                    try:
+                        with io.open(self.path_edge(pin), 'w+b'):
+                            pass
+                    except IOError as e:
+                        if e.errno == errno.EACCES:
+                            sleep(i / 100)
+                        else:
+                            raise
+                if result is None:
                     raise RuntimeError('failed to export pin %d' % pin)
-            else:
-                result = io.open(self.path('gpio%d/value' % pin), 'w+b', buffering=0)
-            self._pin_refs[pin] += 1
             return result
 
     def unexport(self, pin):
         with self._lock:
-            self._pin_refs[pin] -= 1
-            if self._pin_refs[pin] == 0:
-                with io.open(self.path('unexport'), 'wb') as f:
-                    f.write(str(pin).encode('ascii'))
+            try:
+                os.close(self._exports.pop(pin))
+            except KeyError:
+                # unexport should be idempotent
+                pass
+            else:
+                try:
+                    with io.open(self.path('unexport'), 'w+b') as f:
+                        f.write(str(pin).encode('ascii'))
```

Comment:
Is this assuming that by the time the `export` file is writeable, then the `unexport` file must also be writeable at the same time?